# 02 EDA — Customers

Read-only inspection of `olist_customers_dataset.csv`. No cleaning, charts, or feature engineering.

## 1. Import libraries

Import pandas for tabular inspection.

In [1]:
import pandas as pd

**Observation:** pandas is ready.

## 2. Load CSV

Load only this CSV into memory without changing the source.

In [2]:
file_path = '../data/raw/olist/olist_customers_dataset.csv'
df = pd.read_csv(file_path)

**Observation:** The table is loaded; the raw CSV is unchanged.

## 3. Shape

Check table size.

In [3]:
df.shape

(99441, 5)

**Observation:** This reports the table's row and column count.

## 4. First 5 rows

Preview the raw structure.

In [4]:
df.head()

                        customer_id  ... customer_state
0  06b8999e2fba1a1fbc88172c00ba8bc7  ...             SP
1  18955e83d337fd6b2def6b18a428ac77  ...             SP
2  4e7b3e00288586ebd08712fdd0374a03  ...             SP
3  b2b6027bc5c5109e529d4dc6358b12c3  ...             SP
4  4f2d8ab171c80ec8364f7c12e35b23ad  ...             SP

[5 rows x 5 columns]

**Observation:** The preview confirms the row grain and field layout.

## 5. Column names

List available fields.

In [5]:
df.columns.tolist()

['customer_id', 'customer_unique_id', 'customer_zip_code_prefix', 'customer_city', 'customer_state']

**Observation:** These are the fields available for this table's analysis.

## 6. Data types

Review pandas data types.

In [6]:
df.dtypes

customer_id                   str
customer_unique_id            str
customer_zip_code_prefix    int64
customer_city                 str
customer_state                str
dtype: object

**Observation:** Types show which fields are text, numeric, or converted datetime values.

## 7. Missing values

Count nulls by field.

In [7]:
df.isna().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

**Observation:** Missing counts identify incomplete fields without changing them.

## 8. Missing percentage

Measure missingness relative to table size.

In [8]:
(df.isna().mean().mul(100).round(2)).sort_values(ascending=False)

customer_id                 0.0
customer_unique_id          0.0
customer_zip_code_prefix    0.0
customer_city               0.0
customer_state              0.0
dtype: float64

**Observation:** Percentages show the relative scale of incomplete fields.

## 9. Duplicate rows

Count complete row duplicates.

In [9]:
df.duplicated().sum()

np.int64(0)

**Observation:** This is the number of repeated full rows; none are removed.

## 10. Duplicate primary keys

Validate the documented candidate key.

In [10]:
primary_key = ['customer_id']
df.duplicated(subset=primary_key, keep=False).sum()

np.int64(0)

**Observation:** This counts rows involved in duplicate candidate keys for customer_id.

## 11. Unique values

Inspect meaningful category or entity counts.

In [11]:
pd.Series({'customer_id': df['customer_id'].nunique(), 'customer_unique_id': df['customer_unique_id'].nunique(), 'customer_states': df['customer_state'].nunique(), 'customer_cities': df['customer_city'].nunique()})

customer_id           99441
customer_unique_id    96096
customer_states          27
customer_cities        4119
dtype: int64

**Observation:** This gives a concise view of important distinct values.

## 12. Basic describe()

Generate a basic summary for all fields.

In [12]:
df.describe(include='all')

                             customer_id  ... customer_state
count                              99441  ...          99441
unique                             99441  ...             27
top     06b8999e2fba1a1fbc88172c00ba8bc7  ...             SP
freq                                   1  ...          41746
mean                                 NaN  ...            NaN
std                                  NaN  ...            NaN
min                                  NaN  ...            NaN
25%                                  NaN  ...            NaN
50%                                  NaN  ...            NaN
75%                                  NaN  ...            NaN
max                                  NaN  ...            NaN

[11 rows x 5 columns]

**Observation:** The summary describes completeness, distinctness, common values, and numeric ranges where applicable.

## 13. Table-specific observations

The table connects 99,441 order-level customer IDs to 96,096 stable customer identities. State and city fields support future geographic analysis.

## 14. Data quality observations

No missing or exact duplicate rows were found, and `customer_id` is unique. `customer_unique_id` repeats intentionally for repeat buyers; some ZIP prefixes lack geolocation matches, as summarized in the final notebook.